### PR TITLE
Use C++ templates to represent Rust generics

### LINKED ISSUE #85: Avoid monomorphisation for C++?
I assume I'm missing something obvious, but is it really necessary to run monomorphisation for C++ output instead of converting generics to `template`s?

Are there cases where output and/or layout would be different from such in Rust?

### PR DESCRIPTION
**Pull request by RReverser**
This implements native representation of Rust generics as templates when generating C++ code.

Parameter for generic template specialisation which generated template groups on top of monomorphised structs was removed as it's now covered by just regular generic templates.

Didn't touch C part yet as it's somewhat more involved, TODO later.

Fixes #85 

**UPD**: Also added generic typedef support instead of manual specialization, which makes aliased types compatible with their unaliased counterparts in C++ (as they resolve to the same type) and significantly simplifies implementation.

### CODE REVIEW

**Comment by RReverser** [comment]
@eqrion What do you think in general? I'll be happy to do C part too, but would love to hear any concerns before I got too far just in case this turns out unacceptable.

**Comment by RReverser** [comment]
(also, complexity of generic part of #78 in other branch depends on whether this will be accepted approach or not, as it simplifies implementing new generics)

**Comment by eqrion** [comment]
This looks great to me. Thanks for doing this!

I will need to do some investigation tomorrow to ensure this doesn't cause any unreasonable breakage for [webrender](https://github.com/servo/webrender), but after a first pass this looks good.

**Comment by RReverser** [comment]
Sure, thank you for reviews!

**Comment by RReverser** [comment]
> If you're interested the diff is here. I'm happy to see mangled names disappear.

That went surprisingly smoothly :)

File: src/bindgen/bindings.rs [hunk 1]
```diff
@@ -123,6 +120,12 @@ impl Bindings {
             out.new_line();
         }
 
+        if self.config.language == Language::Cxx {
+            out.new_line_if_not_start();
+            out.write("extern \"C\" {");
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/bindgen/bindings.rs]: Oh yeah, sure.

[the remark this answers is not in the document]
Comment (RReverser) [reply src/bindgen/bindings.rs]: It would be nice to have tests covering these nuances :) Done.

**Comment by eqrion** [comment]
Thanks!

**Pull request merged**